### PR TITLE
Parses empty unquote Xml expression

### PR DESCRIPTION
**Pull request by allanrenucci**
Parses "unquote Xml expression" similarly to "unquote expression".
This fixes a bug with empty expression (e.g. <a>{}</a>).

### CODE REVIEW

File: scalameta/scalameta/shared/src/test/scala/scala/meta/tests/parsers/XmlSuite.scala [hunk 1]
```diff
@@ -288,7 +288,7 @@ class XmlSuite extends ParseSuite {
     """.trim.stripMargin)
 
 
-  //checkOK("<a>{}</a>") // FIXME: Should parse
+  checkOK("<a>{}</a>")
```
Comment (xeno-by) [inline scalameta/scalameta/shared/src/test/scala/scala/meta/tests/parsers/XmlSuite.scala]: What does this code mean?

Comment (allanrenucci) [inline scalameta/scalameta/shared/src/test/scala/scala/meta/tests/parsers/XmlSuite.scala]: scalac insert `()`: `<a>{(): Unit}</a>`. Same for string interpolation `s"${}"`

Comment (xeno-by) [inline scalameta/scalameta/shared/src/test/scala/scala/meta/tests/parsers/XmlSuite.scala]: Wow this is crazy 😮

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 2]
```diff
@@ -1591,7 +1591,7 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
 
   def unquoteXmlExpr(): Term = {
     // TODO: verify this
-    dropAnyBraces(expr())
+    unquoteExpr()
   }
```
Comment (xeno-by) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala]: What about other usages of `dropAnyBraces`, like in `unquoteXmlPattern`?

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 3]
```diff
@@ -1591,7 +1591,7 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
 
   def unquoteXmlExpr(): Term = {
     // TODO: verify this
```
Comment (xeno-by) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala]: I guess this comment can be deleted now.

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala (hunk 2 above)
Comment (allanrenucci) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala]: I haven't had a look to xml in pattern position. Seems to be very similar to interpolator in pattern position. Will need to investigate more.

**Comment by allanrenucci** [comment]
@xeno-by @olafurpg How should `s"${}"` be parsed?
scalac doesn't differentiate between `s"${}"` and`s"${()}"` but scalameta does:

scalareflect:
```scala
> showRaw(q""" s"$${}" """) 
Apply(Select(Apply(Ident(TermName("StringContext")), List(Literal(Constant("")), Literal(Constant("")))), TermName("s")), List(Literal(Constant(()))))
> showRaw(q""" s"$${()}" """) 
Apply(Select(Apply(Ident(TermName("StringContext")), List(Literal(Constant("")), Literal(Constant("")))), TermName("s")), List(Literal(Constant(()))))
```
scalameta:
```scala
> s"""s"Hello $${}"""".parse[Stat].get.structure
Term.Interpolate(Term.Name("s"), Seq(Lit.String(""), Lit.String("")), Seq(Term.Block(Nil)))
> s"""s"Hello $${()}"""".parse[Stat].get.structure
Term.Interpolate(Term.Name("s"), Seq(Lit.String(""), Lit.String("")), Seq(Lit.Unit(())))
```

I think we should parse xml literals the same way we parse string interpolators (modulo mandatory `{}` for xml literals). This is what scalac does.

**Comment by olafurpg** [comment]
I propose that we do the same thing in xml literals as we do in string interpolators in scalameta, so the parser should emit `Term.Block(Nil)`. Either that, or we insert `Lit.Unit(())` in both to be the same as scalac. I agree that xml and regular interpolators should behave the same in this case. However, the scalac behavior of inserting a `Lit.Unit(())` is just baffling to me.

**Comment by allanrenucci** [comment]
@olafurpg This is what happen in the scalac parser:
```scala
def block(): Tree = makeBlock(blockStatSeq())

def makeBlock(stats: List[Tree]): Tree = gen.mkBlock(stats)

def mkBlock(stats: List[Tree], doFlatten: Boolean = true): Tree =
  if (stats.isEmpty) mkSyntheticUnit()
  else if (!stats.last.isTerm) Block(stats, mkSyntheticUnit())
  else if (stats.length == 1 && doFlatten) stats.head
  else Block(stats.init, stats.last)
```
So when it parses a an empty block, it inserts a synthetic unit.

**Comment by olafurpg** [comment]
I'm OK with reproducing what the scalac parser does, wdyt @xeno-by ?

**Comment by xeno-by** [comment]
I think we should do `Term.Block(Nil)`. Scalac does a lot of desugarings, but we're not obliged to, and to me `{}` looks like an empty block, hence `Term.Block(Nil)`. Wdyt?

**Comment by olafurpg** [comment]
OK, `Term.Block(Nil)` it is then!

**Comment by allanrenucci** [comment]
Then I guess it is good to go

**Comment by allanrenucci** [comment]
>t's interesting to note that the Syntax is {{}}, since the difference between {{}} and {} is now considered token trivia.

@olafurpg I have:
```scala
tree.show[Syntax] == "<foo>{}</foo>"
```

File: scalameta/scalameta/shared/src/test/scala/scala/meta/tests/parsers/XmlSuite.scala [hunk 4]
```diff
@@ -198,6 +198,12 @@ class XmlSuite extends ParseSuite {
     // format: on
   }
 
+  test("xml literals unit") {
+    val tree = term("<foo>{}</foo>")
+    assert(tree.show[Structure] == """Term.Xml(Seq(Lit.String("<foo>"), Lit.String("</foo>")), Seq(Term.Block(Nil)))""")
+    assert(tree.show[Syntax] == "<foo>{}</foo>")
```
Comment (olafurpg) [inline scalameta/scalameta/shared/src/test/scala/scala/meta/tests/parsers/XmlSuite.scala]: We need to remove the original tokens as is done here https://github.com/scalameta/scalameta/blob/2128ff7b233aafb797c8c1b9bc747a67e10efab4/scalameta/scalameta/shared/src/test/scala/scala/meta/tests/prettyprinters/SyntacticSuite.scala#L22 I propose we move this test case to SyntacticSuite right under the `xml literals` test.

Comment (allanrenucci) [inline scalameta/scalameta/shared/src/test/scala/scala/meta/tests/parsers/XmlSuite.scala]: Done

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala (hunk 2 above)
Comment (xeno-by) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala]: Did you have a chance to look into this further?

Comment (allanrenucci) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala]: If I remove `dropAnyBraces` in `unquoteXmlPattern` then it fails to parse xml literals in pattern position.
Also I added a test to make sure that empty splices in pattern position fail to parse (like in scalac):
```scala
checkError("e match { case <a>{}</a> => ??? }")
```

Comment (xeno-by) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala]: Thanks a lot for your investigation!
